src/debug.c

```diff
@@ -0,0 +1,603 @@
+extern int8_t usb_64drive_wait(void);
+extern void usb_64drive_setwritable(int8_t enable);
+
+static DRESULT fat_disk_read_64drive(BYTE* buff, LBA_t sector, UINT count)
+{
+	_Static_assert(FF_MIN_SS == 512, "this function assumes sector size == 512");
+	_Static_assert(FF_MAX_SS == 512, "this function assumes sector size == 512");
+
+	for (int i=0;i<count;i++)
+	{	
+		usb_64drive_wait();
+		io_write(D64_CIBASE_ADDRESS + D64_REGISTER_LBA, sector+i);
+		usb_64drive_wait();
+		io_write(D64_CIBASE_ADDRESS + D64_REGISTER_COMMAND, D64_COMMAND_SD_READ);
+		if (usb_64drive_wait() != 0)
+		{
+			// Operation is taking too long. Probably SD was not inserted.
+			// Send a COMMAND_ABORT and SD_RESET, and return I/O error.
+			// Note that because of a 64drive firmware bug, this is not
+			// sufficient to unblock the 64drive. The USB channel will stay
+			// unresponsive. We don't currently have a workaround for this.
+			io_write(D64_CIBASE_ADDRESS + D64_REGISTER_COMMAND, D64_COMMAND_ABORT);
+			usb_64drive_wait();
+			io_write(D64_CIBASE_ADDRESS + D64_REGISTER_COMMAND, D64_COMMAND_SD_RESET);
+			usb_64drive_wait();
+			return FR_DISK_ERR;
+		}
+
+		if (((uint32_t)buff & 7) == 0) 
+		{
+			data_cache_hit_writeback_invalidate(buff, 512);
+			dma_read(buff, D64_CIBASE_ADDRESS + D64_BUFFER, 512);
+		}
+		else
+		{
+			typedef uint32_t u_uint32_t __attribute__((aligned(1)));
+
+			uint32_t* src = (uint32_t*)(D64_CIBASE_ADDRESS + D64_BUFFER);
+			u_uint32_t* dst = (u_uint32_t*)buff;
+			for (int i = 0; i < 512/16; i++) 
+			{
+				uint32_t a = *src++; uint32_t b = *src++; uint32_t c = *src++; uint32_t d = *src++;
+				*dst++ = a;          *dst++ = b;          *dst++ = c;          *dst++ = d;
+			}
+		}
+		buff += 512;
+	}
+	return RES_OK;
+}
+
+static DRESULT fat_disk_write_64drive(const BYTE* buff, LBA_t sector, UINT count)
+{
+	_Static_assert(FF_MIN_SS == 512, "this function assumes sector size == 512");
+	_Static_assert(FF_MAX_SS == 512, "this function assumes sector size == 512");
+
+	for (int i=0;i<count;i++)
+	{	
+		if (((uint32_t)buff & 7) == 0) 
+		{
+			data_cache_hit_writeback(buff, 512);
+			dma_write(buff, D64_CIBASE_ADDRESS + D64_BUFFER, 512);
+		}
+		else
+		{
+			typedef uint32_t u_uint32_t __attribute__((aligned(1)));
+
+			uint32_t* dst = (uint32_t*)(D64_CIBASE_ADDRESS + D64_BUFFER);
+			u_uint32_t* src = (u_uint32_t*)buff;
+			for (int i = 0; i < 512/16; i++) 
+			{
+				uint32_t a = *src++; uint32_t b = *src++; uint32_t c = *src++; uint32_t d = *src++;
+				*dst++ = a;          *dst++ = b;          *dst++ = c;          *dst++ = d;
+			}
+		}
+
+		usb_64drive_wait();
+		io_write(D64_CIBASE_ADDRESS + D64_REGISTER_LBA, sector+i);
+		usb_64drive_wait();
+		io_write(D64_CIBASE_ADDRESS + D64_REGISTER_COMMAND, D64_COMMAND_SD_WRITE);
+		usb_64drive_wait();
+	}
+
+	return RES_OK;
+}
+
+static DRESULT fat_disk_ioctl_64drive(BYTE cmd, void* buff)
+{
+	switch (cmd)
+	{
+	case CTRL_SYNC: return RES_OK;
```

Comment:
I think as a followup optimization, we can try to remove the very last `usb_64drive_wait()` from `fat_disk_write_64drive` and add it to `CTRL_SYNC`. Notice that this would parallelize only the last sector write in a group of write performed by `fat_disk_write_64drive`, so maybe it's not really worth it. To make something more elaborated, we would need a background thread to perform writes, but that sounds complicated, and also I'm not sure you can do that so easily -- the contents of `buff` might not be available anymore after `fat_disk_write_64drive` returns. 

At the end of the day, it's probably better to leave it synchronous like this, and then ask the user to move sdfs accesses in a different thread if they wish so. This would still require that `usb_64drive_wait` yields, which is complicated by the fact that we don't get an interrupt from 64drive or ED64 when the command is finished. Sorry, not many good news here :)